File: src/domain/models/preview.py
"""Value objects for the bootstrap preview.

FileAction, GlobalSettingConflict, ConflictResolution, and Preview are
immutable value objects that represent what alty *will* do before it does it.
The preview-before-action invariant guarantees users see every change first.

Design decisions:
- FileActionType has no OVERWRITE variant. Alty never silently overwrites.
- Conflict rename suffix is ``_alty``.
- All value objects are frozen dataclasses (stdlib only, no Pydantic).
"""

from __future__ import annotations

import enum
from dataclasses import dataclass
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from pathlib import Path


class FileActionType(enum.Enum):
    """The kind of file operation alty plans to perform.

    CREATE:          Write a new file that does not exist yet.
    SKIP:            Leave an existing file untouched.
    CONFLICT_RENAME: Rename an existing file with the ``_alty`` suffix
                     before writing the template version.
    """

    CREATE = "create"
    SKIP = "skip"
    CONFLICT_RENAME = "conflict_rename"


@dataclass(frozen=True)
class FileAction:
    """A single planned file operation.

    Attributes:
        path: Relative path from the project root.
        action_type: What alty will do with this file.
        reason: Human-readable explanation (e.g. "already exists").
        renamed_path: Only set for CONFLICT_RENAME actions.
    """

    path: Path
    action_type: FileActionType
    reason: str = ""
    renamed_path: Path | None = None


@dataclass(frozen=True)
class GlobalSettingConflict:
    """A conflict between a global tool setting and the local project config.

    Attributes:
        tool: Tool identifier (e.g. "cursor", "claude").
        global_path: Absolute path to the global settings file.
        global_value: Current global value as a string.
        local_value: Value alty wants to set locally.
    """

    tool: str
    global_path: Path
    global_value: str
    local_value: str


class ConflictResolution(enum.Enum):
    """How to resolve a GlobalSettingConflict.

    KEEP_GLOBAL:            Leave the global setting and skip local config.
    UPDATE_GLOBAL:          Update the global setting to the new value.
    SET_LOCAL_WITH_WARNING: Write local config with a warning comment.
    """

    KEEP_GLOBAL = "keep_global"
    UPDATE_GLOBAL = "update_global"
    SET_LOCAL_WITH_WARNING = "set_local_with_warning"


@dataclass(frozen=True)
class Preview:
    """Immutable snapshot of all planned bootstrap actions.

    Attributes:
        file_actions: Tuple of planned file operations.
        conflicts: Tuple of detected global-setting conflicts (may be empty).
    """

    file_actions: tuple[FileAction, ...]
    conflicts: tuple[GlobalSettingConflict, ...] = ()
    conflict_descriptions: tuple[str, ...] = ()

File: internal/shared/domain/valueobjects/preview.go
package valueobjects

// FileActionType is the kind of file operation alto plans to perform.
type FileActionType string

// File action type constants.
const (
	FileActionCreate         FileActionType = "create"
	FileActionSkip           FileActionType = "skip"
	FileActionConflictRename FileActionType = "conflict_rename"
)

// AllFileActionTypes returns all valid file action type values.
func AllFileActionTypes() []FileActionType {
	return []FileActionType{FileActionCreate, FileActionSkip, FileActionConflictRename}
}

// FileAction is a single planned file operation.
type FileAction struct {
	path        string
	actionType  FileActionType
	reason      string
	renamedPath string
}

// NewFileAction creates a FileAction value object.
func NewFileAction(path string, actionType FileActionType, reason string, renamedPath string) FileAction {
	return FileAction{
		path:        path,
		actionType:  actionType,
		reason:      reason,
		renamedPath: renamedPath,
	}
}

// Path returns the file path.
func (fa FileAction) Path() string { return fa.path }

// ActionType returns the action type.
func (fa FileAction) ActionType() FileActionType { return fa.actionType }

// Reason returns the human-readable reason.
func (fa FileAction) Reason() string { return fa.reason }

// RenamedPath returns the renamed path (empty if not a conflict rename).
func (fa FileAction) RenamedPath() string { return fa.renamedPath }

// GlobalSettingConflict is a conflict between a global tool setting and local project config.
type GlobalSettingConflict struct {
	tool        string
	globalPath  string
	globalValue string
	localValue  string
}

// NewGlobalSettingConflict creates a GlobalSettingConflict value object.
func NewGlobalSettingConflict(tool, globalPath, globalValue, localValue string) GlobalSettingConflict {
	return GlobalSettingConflict{
		tool:        tool,
		globalPath:  globalPath,
		globalValue: globalValue,
		localValue:  localValue,
	}
}

// Tool returns the tool identifier.
func (c GlobalSettingConflict) Tool() string { return c.tool }

// GlobalPath returns the path to the global settings file.
func (c GlobalSettingConflict) GlobalPath() string { return c.globalPath }

// GlobalValue returns the current global value.
func (c GlobalSettingConflict) GlobalValue() string { return c.globalValue }

// LocalValue returns the desired local value.
func (c GlobalSettingConflict) LocalValue() string { return c.localValue }

// ConflictResolution describes how to resolve a GlobalSettingConflict.
type ConflictResolution string

// Conflict resolution constants.
const (
	ConflictKeepGlobal          ConflictResolution = "keep_global"
	ConflictUpdateGlobal        ConflictResolution = "update_global"
	ConflictSetLocalWithWarning ConflictResolution = "set_local_with_warning"
)

// Preview is an immutable snapshot of all planned bootstrap actions.
type Preview struct {
	fileActions          []FileAction
	conflicts            []GlobalSettingConflict
	conflictDescriptions []string
}

// NewPreview creates a Preview value object.
func NewPreview(fileActions []FileAction, conflicts []GlobalSettingConflict, conflictDescriptions []string) Preview {
	fa := make([]FileAction, len(fileActions))
	copy(fa, fileActions)
	c := make([]GlobalSettingConflict, len(conflicts))
	copy(c, conflicts)
	cd := make([]string, len(conflictDescriptions))
	copy(cd, conflictDescriptions)
	return Preview{
		fileActions:          fa,
		conflicts:            c,
		conflictDescriptions: cd,
	}
}

// FileActions returns a defensive copy of file actions.
func (p Preview) FileActions() []FileAction {
	out := make([]FileAction, len(p.fileActions))
	copy(out, p.fileActions)
	return out
}

// Conflicts returns a defensive copy of conflicts.
func (p Preview) Conflicts() []GlobalSettingConflict {
	out := make([]GlobalSettingConflict, len(p.conflicts))
	copy(out, p.conflicts)
	return out
}

// ConflictDescriptions returns a defensive copy of conflict descriptions.
func (p Preview) ConflictDescriptions() []string {
	out := make([]string, len(p.conflictDescriptions))
	copy(out, p.conflictDescriptions)
	return out
}
